From cf3b1439547dce7fd8fc4d113c2d3b6cff273716 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Tue, 14 Nov 2017 02:47:51 +0100
Subject: [PATCH 1057/1089] ARM: dts: axp813: Enable power supply led control

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 arch/arm/boot/dts/axp81x.dtsi | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/arm/boot/dts/axp81x.dtsi b/arch/arm/boot/dts/axp81x.dtsi
index f32a8ba53b85..6f4ea939b6ea 100644
--- a/arch/arm/boot/dts/axp81x.dtsi
+++ b/arch/arm/boot/dts/axp81x.dtsi
@@ -170,4 +170,8 @@
 	usb_power_supply: usb-power-supply {
 		compatible = "x-powers,axp813-usb-power-supply";
 	};
+
+	power_supply_leds: power-supply-leds {
+		compatible = "x-powers,axp813-leds";
+	};
 };
-- 
2.19.2

